src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs

```diff
@@ -2556,13 +2574,14 @@ internal override NamedTypeSymbol TryLookupForwardedMetadataTypeWithCycleDetecti
             if (_lazyForwardedTypesFromSource == null)
             {
                 IDictionary<string, NamedTypeSymbol> forwardedTypesFromSource;
-                CommonAssemblyWellKnownAttributeData<NamedTypeSymbol> wellKnownAttributeData = GetSourceDecodedWellKnownAttributeData();
+                // Do minimal binding of attributes (just ForwardedTypes) to avoid cycle problems
```

Comment:
>// Do minimal binding of attributes (just ForwardedTypes) to avoid cycle problems [](start = 16, length = 81)

If we already bound all attributes from source, it probably makes sense to use information from that rather than to rebind/reanalyze attributes again.

Reply:
Good idea. Thanks
I couldn't find a way to test that path though

---
In reply to: [155961618](https://github.com/dotnet/roslyn/pull/23647#discussion_r155961618) [](ancestors = 155961618)

Reply:
> I couldn't find a way to test that path though

We should have a test that verifies that type forwarders are handled properly after GetAttributes is called for the assembly.

---
In reply to: [156215452](https://github.com/dotnet/roslyn/pull/23647#discussion_r156215452) [](ancestors = 156215452,155961618)

Reply:
Thanks for the suggestion. I added this to one of tests, but still didn't manage to hit the branch in `GetForwardedTypes` where we used cached attributes.
After `GetAttributes` was called, calling `VerifyDiagnostics` no longer even calls `TryLookupForwardedMetadataTypeWithCycleDetection`. I'm not sure it's worthwhile to investigate where the executions diverge (probably some other caching upstream).

---
In reply to: [156455955](https://github.com/dotnet/roslyn/pull/23647#discussion_r156455955) [](ancestors = 156455955,156215452,155961618)

Reply:
The scenario you explained indeed worked and hit the desired path. Thanks!

---
In reply to: [156468659](https://github.com/dotnet/roslyn/pull/23647#discussion_r156468659) [](ancestors = 156468659,156455955,156215452,155961618)